// node-sp-auth-config/dist/config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const spauth = require("node-sp-auth");
exports.getTargetsTypes = () => {
    return ['Online', 'OnPremise'];
};
exports.getStrategies = () => {
    const strategies = [
        {
            id: 'OnpremiseUserCredentials',
            name: 'User credentials (NTLM)',
            withPassword: true,
            target: ['OnPremise'],
            verifyCallback: spauth.isUserCredentialsOnpremise
        }, {
            id: 'OnpremiseFbaCredentials',
            name: 'Form-based authentication (FBA)',
            withPassword: true,
            target: ['OnPremise'],
            verifyCallback: spauth.isFbaCredentialsOnpremise
        }, {
            id: 'OnpremiseTmgCredentials',
            name: 'Forefront TMG authentication',
            withPassword: true,
            target: ['OnPremise'],
            verifyCallback: spauth.isTmgCredentialsOnpremise
        }, {
            id: 'OnPremiseAddinCredentials',
            name: 'Add-In only permissions',
            withPassword: false,
            target: ['OnPremise'],
            verifyCallback: (...args) => {
                return spauth.isAddinOnlyOnpremise(args[1]);
            }
        }, {
            id: 'UserCredentials',
            name: 'User credentials (SAML)',
            withPassword: true,
            target: ['Online'],
            verifyCallback: spauth.isUserCredentialsOnline
        }, {
            id: 'OnlineAddinCredentials',
            name: 'Add-In only permissions',
            withPassword: true,
            target: ['Online'],
            verifyCallback: (...args) => {
                return spauth.isAddinOnlyOnline(args[1]);
            }
        }, {
            id: 'AdfsUserCredentials',
            name: 'ADFS user credentials',
            withPassword: true,
            target: ['Online', 'OnPremise'],
            verifyCallback: (...args) => {
                return spauth.isAdfsCredentials(args[1]);
            }
        }, {
            id: 'OnDemandCredentials',
            name: 'On-demand credentials',
            withPassword: false,
            target: ['Online', 'OnPremise'],
            verifyCallback: (...args) => {
                return spauth.isOndemandCredentials(args[1]);
            }
        }
    ];
    return strategies;
};
